# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Basics

### Exercise 1: Importing and Inspecting your own Dataset

Before we can work with the Pandas Library, we need to __import Pandas__ (by convention as __pd__). <br>
1. __Import__ Pandas. __Fill in the gaps__!

In [1]:
import pandas as pd

Now we want to import the __Listings Dataset__ that is stored in the CSV-File __listings.csv__. <br>
2. __Import__ the Listings Dataset and __store__ the Pandas DataFrame in the variable __listings__! __Fill in the gaps__!

In [2]:
listings = pd.read_csv("listings.csv")

#### Some more details on the Columns:

- __Symbol__: A Stock Ticker Symbol is an abbreviation used to uniquely identify publicly traded shares of a particular stock on a particular stock market
- __Exchange__: Marketplace in which shares are traded 
- __Name__: Legal Company Name
- __Last_Price__: Last Trading Price (as of April 2019)
- __Market_Cap__: Dollar Value of Outstanding Shares (as of April 2019). Computed as Shares times current Price. 
- __ADR_TSO__: Additional Information on foreign stocks trading in the US 
- __IPO_Year__: Year of Initial Public Offering 
- __Sector__: Sector of main business activity
- __Industry__: Industry of main business activity


3. __Inspect__ the __first 10 Rows__ of the DataFrame listings! __Fill in the gaps__! <br> What is the __company name of the very first listing__ in the first row (@ index label/pos = 0)?

In [3]:
listings.head(10)

Symbol Exchange                              Name  Last_Price    Market_Cap  \
0      A     NYSE        Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE                 Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                       Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE                AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ     American Airlines Group, Inc.       34.02  1.527687e+10   
5   AAMC     AMEX  Altisource Asset Management Corp       29.90  4.738157e+07   
6   AAME   NASDAQ     Atlantic American Corporation        2.48  4.998398e+07   
7    AAN     NYSE                Aaron&#39;s,  Inc.       53.54  3.628838e+09   
8   AAOI   NASDAQ     Applied Optoelectronics, Inc.       12.30  2.445561e+08   
9   AAON   NASDAQ                        AAON, Inc.       44.88  2.336955e+09   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   
5      NaN       NaN           Finance   
6      NaN       NaN           Finance   
7      NaN       NaN        Technology   
8      NaN    2013.0        Technology   
9      NaN       NaN     Capital Goods   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services  
5                                       Real Estate  
6                                    Life Insurance  
7                   Diversified Commercial Services  
8                                    Semiconductors  
9                   Industrial Machinery/Components

It´s ... Agilent Technologies, Inc.!

4. Also __inspect__ the __last 5 Rows__! __How many listings/rows__ do we have in the Dataset (mind the range index starting from 0!)?

In [4]:
listings.tail()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
6847   ZUMZ   NASDAQ                    Zumiez Inc.       26.72  6.819191e+08   
6848    ZUO     NYSE                    Zuora, Inc.       19.79  2.147266e+09   
6849   ZYME     NYSE                 Zymeworks Inc.       15.74  5.040782e+08   
6850   ZYNE   NASDAQ  Zynerba Pharmaceuticals, Inc.        7.85  1.653995e+08   
6851   ZYXI   NASDAQ                    Zynex, Inc.        5.02  1.618349e+08   

      ADR TSO  IPO_Year             Sector  \
6847      NaN    2005.0  Consumer Services   
6848      NaN    2018.0         Technology   
6849      NaN    2017.0        Health Care   
6850      NaN    2015.0        Health Care   
6851      NaN       NaN        Health Care   

                                               Industry  
6847                     Clothing/Shoe/Accessory Stores  
6848            Computer Software: Prepackaged Software  
6849                              Major Pharmaceuticals  
6850                              Major Pharmaceuticals  
6851  Biotechnology: Electromedical & Electrotherape...

We have ... 6852 listings!

5. Get some __meta information__ on our DataFrame! In which __Column(s)__ do we have __missing/NaN__ Values?

In [5]:
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 9 columns):
Symbol        6852 non-null object
Exchange      6852 non-null object
Name          6852 non-null object
Last_Price    6745 non-null float64
Market_Cap    5954 non-null float64
ADR TSO       140 non-null float64
IPO_Year      3105 non-null float64
Sector        5309 non-null object
Industry      5309 non-null object
dtypes: float64(4), object(5)
memory usage: 481.9+ KB


We have missing/NaN Values in the columns ... Last_Price, Market_Cap, ADR TSO, IPO_Year, Sector & Industry!

6. Let´s get some __summary statistics__ on our DataFrame! What is the __maximum__ value in the column __Last_Price__?

In [6]:
listings.describe()

Last_Price    Market_Cap       ADR TSO     IPO_Year
count  6745.000000  5.954000e+03  1.400000e+02  3105.000000
mean     37.372232  7.174440e+09  5.869242e+07  2009.025765
std      99.249040  3.525455e+10  1.629909e+08     9.337562
min       0.003000  0.000000e+00  2.400000e+04  1972.000000
25%       8.310000  1.203769e+08  4.271224e+06  2003.000000
50%      20.573500  5.585176e+08  1.406964e+07  2013.000000
75%      37.370000  2.888747e+09  3.962613e+07  2017.000000
max    4299.990000  9.459795e+11  1.485426e+09  2019.000000

The maximum value in the column Last_Price is ... 4299.99!

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Hints (Spoiler!)

1. pandas as pd

2. Pass the filename in quotation marks to pd.read_csv()

3. head() method with n = 10

4. tail() method

5. info() method

6. describe() method